fabtests/prov/efa/src/rdm_ep_rnr_retry.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) 2021, Amazon.com, Inc.  All rights reserved.
+ *
+ * This software is available to you under the BSD license
+ * below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * This program tests the RNR retry counter reset via fi_setopt.
+ * When running the test, use `-R` option to specify RNR retry counter.
+ * The valid values are 0 - 7 (7 inidicates infinite retry on firmware).
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <shared.h>
+#include <rdma/fi_cm.h>
+#include <rdma/fi_ext.h>
+
+static int poll_rnr_cq_error(void)
+{
+	struct fi_cq_data_entry comp;
+	struct fi_cq_err_entry comp_err;
+	int total_send, expected_rnr_error;
+	int ret, i, cnt, rnr_flag;
+
+	rnr_flag = 0;
+	expected_rnr_error = 1;
+	/*
+	 * In order for the sender to get RNR error, we need to first consume
+	 * all pre-posted receive buffer (controlled by fi->rx_attr->size)
+	 * on the receiver side, the subsequent sends (expected_rnr_error)
+	 * will then get RNR errors.
+	 */
+	total_send = fi->rx_attr->size + expected_rnr_error;
+
+	for (cnt = total_send, i = 0; i < total_send; i++) {
```

Comment:
cnt is not used in this loop, I prefer to initialize before the next loop, where it is used